README.md
#object.assign

<!--
[![Build Status][travis-svg]][travis-url]
[![dev dependency status][dev-deps-svg]][dev-deps-url]
[![License][license-image]][license-url]

[![browser support][testling-png]][testling-url]
-->

An Object.assign shim for ES5-compliant environments (browsers/node.js/io.js). Is applied
only when needed with a few exceptions for non-compliant implementations.

Takes a minimum of 2 arguments: `target` and `source`.
Takes a variable sized list of source arguments - at least 1, as many as you want.
Throws a TypeError if the `target` argument is `null` or `undefined`.

Most common usage:

In node/io.js:

```js
require('object.assign');
```

In a browser:

```html
<script src="object.assign/index.js"></script>
```

## Example

```js
// Multiple sources!
var target = { a: true };
var source1 = { b: true };
var source2 = { c: true };
var sourceN = { n: true };

var expected = {
	a: true,
	b: true,
	c: true,
	n: true
};

require('object.assign');
var assert = require('assert');

Object.assign(target, source1, source2, sourceN);
assert.deepEqual(target, expected); // AWESOME!
```

```js
require('object.assign');
var assert = require('assert');
var target = {
	a: true,
	b: true,
	c: true
};
var source1 = {
	c: false,
	d: false
};
var sourceN = {
	e: false
};

var assigned = Object.assign(target, source1, sourceN);
assert.equal(target, assigned); // returns the target object
assert.deepEqual(assigned, {
	a: true,
	b: true,
	c: false,
	d: false,
	e: false
});
```

```js
var assert = require('assert');
/* when Object.assign is not present */
delete Object.assign;
require('object.assign');
assert.equal(typeof Object.assign, "function");

var target = {
	a: true,
	b: true,
	c: true
};
var source = {
	c: false,
	d: false,
	e: false
};

var assigned = assign(target, source);
assert.deepEqual(Object.assign(target, source), assign(target, source));
```

```js
var assert = require('assert');
/* when Object.assign is present */
assert.equal(typeof Object.assign, 'function');
var builtinAssign = Object.assign;
require('object.assign');
assert.equal(builtinAssign, Object.assign);
```

## Tests
Simply clone the repo, `npm install`, and run `npm test`

<!--
[travis-svg]: https://travis-ci.org/ljharb/object.assign.svg
[travis-url]: https://travis-ci.org/ljharb/object.assign
[dev-deps-svg]: https://david-dm.org/ljharb/object.assign/dev-status.svg?theme=shields.io
[dev-deps-url]: https://david-dm.org/ljharb/object.assign#info=devDependencies
[testling-png]: https://ci.testling.com/ljharb/object.assign.png
[testling-url]: https://ci.testling.com/ljharb/object.assign
[license-image]: http://img.shields.io/npm/l/object.assign.svg
[license-url]: LICENSE
-->


CHANGELOG.md
object.assign - 1.0.0 / 2015-04-15
==================================
  * removed all calls to require
  * module now has no dependencies, it now only runs in ES5+ environments
  * removed optional shimming, shim is instead applied when necessary


# ============== Fork ==============

object.assign - 2.0.1 / 2015-04-12
==================================
  * Make sure non-enumerable Symbols are excluded.

object.assign - 2.0.0 / 2015-04-12
==================================
  * Make sure the shim function overwrites a broken implementation with pending exceptions.
  * Ensure shim is not enumerable using `define-properties`
  * Ensure `Object.assign` includes symbols.
  * All grade A-supported `node`/`iojs` versions now ship with an `npm` that understands `^`.
  * Run `travis-ci` tests on `iojs` and `node` v0.12; speed up builds; allow 0.8 failures.
  * Add `npm run security` via `nsp`
  * Update `browserify`, `jscs`, `tape`, `object-keys`, `is`

object.assign - 1.1.1 / 2014-12-14
==================================
  * Actually include the browser build in `npm`

object.assign - 1.1.0 / 2014-12-14
==================================
  * Add `npm run build`, and build an automatic-shimming browser distribution as part of the npm publish process.
  * Update `is`, `jscs`

object.assign - 1.0.3 / 2014-11-29
==================================
  * Revert "optimize --production installs"

object.assign - 1.0.2 / 2014-11-27
==================================
  * Update `jscs`, `is`, `object-keys`, `tape`
  * Add badges to README
  * Name URLs in README
  * Lock `covert` to `v1.0.0`
  * Optimize --production installs

object.assign - 1.0.1 / 2014-08-26
==================================
  * Update `is`, `covert`

object.assign - 1.0.0 / 2014-08-07
==================================
  * Update `object-keys`, `tape`

object.assign - 0.5.0 / 2014-07-31
==================================
  * Object.assign no longer throws on null or undefined sources, per https://bugs.ecmascript.org/show_bug.cgi?id=3096

object.assign - 0.4.3 / 2014-07-30
==================================
  * Don’t modify vars in the function signature, since it deoptimizes v8

object.assign - 0.4.2 / 2014-07-30
==================================
  * Fixing the version number: v0.4.2

object.assign - 0.4.1 / 2014-07-19
==================================
  * Revert "Use the native Object.keys if it’s available."

object.assign - 0.4.0 / 2014-07-19
==================================
  * Use the native Object.keys if it’s available.
  * Fixes [#2](https://github.com/ljharb/object.assign/issues/2).
  * Adding failing tests for [#2](https://github.com/ljharb/object.assign/issues/2).
  * Fix indentation.
  * Adding `npm run lint`
  * Update `tape`, `covert`
  * README: Use SVG badge for Travis [#1](https://github.com/ljharb/object.assign/issues/1) from mathiasbynens/patch-1

object.assign - 0.3.1 / 2014-04-10
==================================
  * Object.assign does partially modify objects if it throws, per https://twitter.com/awbjs/status/454320863093862400

object.assign - 0.3.0 / 2014-04-10
==================================
  * Update with newest ES6 behavior - Object.assign now takes a variable number of source objects.
  * Update `tape`
  * Make sure old and unstable nodes don’t fail Travis

object.assign - 0.2.1 / 2014-03-16
==================================
  * Let object-keys handle the fallback
  * Update dependency badges
  * Adding bower.json

object.assign - 0.2.0 / 2014-03-16
==================================
  * Use a for loop, because ES3 browsers don’t have "reduce"

object.assign - 0.1.1 / 2014-03-14
==================================
  * Updating readme

object.assign - 0.1.0 / 2014-03-14
==================================
  * Initial release.



test/index.js
'use strict';

var test = require('tape');
var origObjectAssign = Object.assign;
delete Object.assign;
require('../index.js');
var assign = Object.assign;
var hasSymbols = typeof Symbol === 'function' && typeof Symbol() === 'symbol';

test('error cases', function (t) {
	t.throws(function () { assign(null); }, TypeError, 'target must be an object');
	t.end();
});

test('non-object sources', function (t) {
	t.deepEqual(assign({ a: 1 }, null, { b: 2 }), { a: 1, b: 2 }, 'ignores null source');
	t.deepEqual(assign({ a: 1 }, { b: 2 }, undefined), { a: 1, b: 2 }, 'ignores undefined source');
	t.end();
});

test('returns the modified target object', function (t) {
	var target = {};
	var returned = assign(target, { a: 1 });
	t.equal(returned, target, 'returned object is the same reference as the target object');
	t.end();
});

test('has the right length', function (t) {
	t.equal(assign.length, 2, 'length is 2 => 2 required arguments');
	t.end();
});

test('merge two objects', function (t) {
	var target = { a: 1 };
	var returned = assign(target, { b: 2 });
	t.deepEqual(returned, { a: 1, b: 2 }, 'returned object has properties from both');
	t.end();
});

test('works with functions', function (t) {
	var target = function () {};
	target.a = 1;
	var returned = assign(target, { b: 2 });
	t.equal(target, returned, 'returned object is target');
	t.equal(returned.a, 1);
	t.equal(returned.b, 2);
	t.end();
});

test('works with primitives', function (t) {
	var target = 2;
	var source = { b: 42 };
	var returned = assign(target, source);
	t.equal(Object.prototype.toString.call(returned), '[object Number]', 'returned is object form of number primitive');
	t.equal(Number(returned), target, 'returned and target have same valueOf');
	t.equal(returned.b, source.b);
	t.end();
});

test('merge N objects', function (t) {
	var target = { a: 1 };
	var source1 = { b: 2 };
	var source2 = { c: 3 };
	var returned = assign(target, source1, source2);
	t.deepEqual(returned, { a: 1, b: 2, c: 3 }, 'returned object has properties from all sources');
	t.end();
});

test('only iterates over own keys', function (t) {
	var Foo = function () {};
	Foo.prototype.bar = true;
	var foo = new Foo();
	foo.baz = true;
	var target = { a: 1 };
	var returned = assign(target, foo);
	t.equal(returned, target, 'returned object is the same reference as the target object');
	t.deepEqual(target, { baz: true, a: 1 }, 'returned object has only own properties from both');
	t.end();
});

test('includes enumerable symbols, after keys', { skip: !hasSymbols }, function (t) {
	var visited = [];
	var obj = {};
	Object.defineProperty(obj, 'a', { get: function () { visited.push('a'); return 42; }, enumerable: true });
	var symbol = Symbol('enumerable');
	Object.defineProperty(obj, symbol, { get: function () { visited.push(symbol); return Infinity; }, enumerable: true });
	var nonEnumSymbol = Symbol('non-enumerable');
	Object.defineProperty(obj, nonEnumSymbol, { get: function () { visited.push(nonEnumSymbol); return -Infinity; }, enumerable: false });
	var target = assign({}, obj);
    t.deepEqual(visited, ['a', symbol], 'key is visited first, then symbol');
    t.equal(target.a, 42, 'target.a is 42');
    t.equal(target[symbol], Infinity, 'target[symbol] is Infinity');
    t.notEqual(target[nonEnumSymbol], -Infinity, 'target[nonEnumSymbol] is not -Infinity');
	t.end();
});

test('does not fail when symbols are not present', function (t) {
	var getSyms;
	if (hasSymbols) {
		getSyms = Object.getOwnPropertySymbols;
		delete Object.getOwnPropertySymbols;
	}

	var visited = [];
	var obj = {};
	Object.defineProperty(obj, 'a', { get: function () { visited.push('a'); return 42; }, enumerable: true });
	if (hasSymbols) {
		var symbol = Symbol();
		Object.defineProperty(obj, symbol, { get: function () { visited.push(symbol); return Infinity; }, enumerable: true });
	}
	var target = assign({}, obj);
    t.equal(target.a, 42, 'target.a is 42');
    t.deepEqual(visited, ['a'], 'only key is visited');

	if (hasSymbols) {
		// sanity check for "visited" array
		t.equal(obj[symbol], Infinity);
		t.deepEqual(visited, ['a', symbol], 'symbol is visited manually');

		Object.getOwnPropertySymbols = getSyms;
	}
	t.end();
});

/*eslint-disable no-new-func */
test('shims the "assign" function on Object ctor', function (t) {
	var reapplyAssignShim = new Function(require('fs').readFileSync(__dirname + '/../index.js', {encoding: 'utf8'}));

	t.test('when Object.assign is present', function (st) {
		var currObjectAssign = Object.assign;
		var pseudoOrigAssign = function () {};
		Object.assign = pseudoOrigAssign;
		reapplyAssignShim();

		st.equal(Object.assign, pseudoOrigAssign, 'Object.assign is not overridden');

		Object.assign = currObjectAssign;
		st.end();
	});

	t.test('when Object.assign is not present', function (st) {
		var currObjectAssign = Object.assign;
		delete Object.assign;
		reapplyAssignShim();

		st.equal(typeof Object.assign, 'function', 'Object.assign is shimmed');
		if (Object.getOwnPropertyDescriptor) {
			st.equal(Object.getOwnPropertyDescriptor(Object, 'assign').enumerable, false, 'is not enumerable');
		}

		Object.assign = currObjectAssign;
		st.end();
	});

	t.test('when Object.assign is present and has pending exceptions', { skip: !origObjectAssign || !Object.preventExtensions }, function (st) {
		// Firefox 37 still has "pending exception" logic in its Object.assign implementation,
		// which is 72% slower than our shim, and Firefox 40's native implementation.
		var thrower = Object.preventExtensions({ 1: 2 });
		var error;
		try { Object.assign(thrower, 'xy'); } catch (e) { error = e; }
		st.equal(error instanceof TypeError, true, 'error is TypeError');
		st.equal(thrower[1], 2, 'thrower[1] === 2');
		st.end();
	});

	t.end();
});
/*eslint-enable no-new-func */

test('working with actual shim', function (t) {
	t.notEqual(Object.assign, origObjectAssign, 'assign shim is not native Object.assign');
	t.end();
});



index.js
(function () {
    'use strict';

    // modified from https://github.com/ljharb/object.assign
    if (Object.assign && Object.preventExtensions) {
        var assignHasPendingExceptions = (function () {
            // Firefox 37 still has "pending exception" logic in its Object.assign implementation,
            // which is 72% slower than our shim, and Firefox 40's native implementation.
            var thrower = Object.preventExtensions({1: 2});
            try {
                Object.assign(thrower, 'xy');
            } catch (e) {
                return thrower[1] === 'y';
            }
        }());
        if (assignHasPendingExceptions) {
            delete Object.assign;
        }
    }

    if (!Object.assign) {
        var keys = Object.keys;
        var defineProperty = Object.defineProperty;
        var canBeObject = function (obj) {
            return typeof obj !== 'undefined' && obj !== null;
        };
        var hasSymbols = typeof Symbol === 'function' && typeof Symbol() === 'symbol';
        var propIsEnumerable = Object.prototype.propertyIsEnumerable;
        var isEnumerableOn = function (obj) {
            return function isEnumerable(prop) {
                return propIsEnumerable.call(obj, prop);
            };
        };

        var assignShim = function assign(target, source1) {
            if (!canBeObject(target)) {
                throw new TypeError('target must be an object');
            }
            var objTarget = Object(target);
            var s, source, i, props;
            for (s = 1; s < arguments.length; ++s) {
                source = Object(arguments[s]);
                props = keys(source);
                if (hasSymbols && Object.getOwnPropertySymbols) {
                    props.push.apply(props, Object.getOwnPropertySymbols(source).filter(isEnumerableOn(source)));
                }
                for (i = 0; i < props.length; ++i) {
                    objTarget[props[i]] = source[props[i]];
                }
            }
            return objTarget;
        };

        defineProperty(Object, 'assign', {
            value: assignShim,
            configurable: true,
            enumerable: false,
            writable: true
        });
    }
}());